File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = 0
        left,right = 0,0
        length = 0
        max_set = set()
        while right < len(s):
            while s[right] in max_set:
                max_set.remove(s[left])
                left+=1
            max_set.add(s[right])
            right+=1
            length = max(length,len(max_set))
        return length

        

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    int maxLen = 0;
public:
    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);
        return maxLen;
    }
private:
    int dfs(TreeNode* root){
        if(!root){return 0;}
        int l = dfs(root->left);
        int r = dfs(root->right);
        maxLen = max(maxLen, l+r);
        return max(l,r)+1;
    }
};
